core/windows_and_tabs/window_layout.py

```diff
@@ -0,0 +1,282 @@
+            )
+    return Gap()
+
+
+def snap_layout(layout: WindowLayout):
+    """Split the screen between multiple windows."""
+    try:
+        global layout_in_progress, last_layout
+        layout_in_progress = layout
+
+        # If called multiple times (and the user hasn't focused a window manually since
+        # last time), rotate the offset of the existing windows in the arrangement,
+        # allowing the user to use a repeater to cycle through the windows to get the
+        # desired result.
+        if (
+            layout.can_rotate
+            and last_layout is not None
+            and last_layout.name == layout.name
+            and layout.windows == last_layout.windows
+        ):
+            layout.rotation_count = last_layout.rotation_count + 1
+
+        # Copy these data structures so we can mutate them:
+        remaining_windows = [w for w in layout.windows]
+        split_positions = layout.split_positions.copy()
+
+        snapped_windows = []
+        for _ in range(layout.rotation_count):
+            split_positions.append(split_positions.pop(0))
+
+        while len(split_positions) > 0:
+            snapped_window: Window = snap_next(
+                remaining_windows, split_positions.pop(0)
+            )
+            snapped_windows.insert(0, snapped_window)
+
+        if len(snapped_windows) > 0:
+            for _ in range(layout.rotation_count):
+                snapped_windows.append(snapped_windows.pop(0))
+
+        for window in snapped_windows:
+            if isinstance(window, Gap):
+                continue
+            actions.user.switcher_focus_window(window)
+
+        layout_in_progress.finish_time = time.perf_counter()
+        last_layout = layout_in_progress
+    finally:
+        layout_in_progress = None
+
+
+def filter_nonviable_windows(windows: List[Window]) -> list[Window]:
+    active_window = ui.active_window()
+
+    # Many invisible non-resizable windows are identifiable because they exist above the current window
+    # in the z-index
+    all_windows = ui.windows()
+    active_window_idx = all_windows.index(active_window)  # type: ignore
+    return list(
+        filter(
+            lambda w: (isinstance(w, Gap) or all_windows.index(w) >= active_window_idx),
+            windows,
+        )
+    )
+
+
+mod = Module()
+mod.list(
+    "window_split_positions",
+    "Predefined window positions when splitting the screen between multiple windows.",
+)
+ctx = Context()
+
+
+@mod.capture(rule="all")
+def all_candidate_windows(m) -> list[Window]:
+    return filter_nonviable_windows(ui.windows())
+
+
+@mod.capture(rule="gap")
+def skip_window(m) -> list[Window]:
+    return [Gap()]
+
+
+@mod.capture(rule="<user.running_applications>")
+def application_windows(m) -> list[Window]:
+    return [
+        window
+        for app in m.running_applications_list
+        for window in actions.self.get_running_app(app).windows()
```

Comment:
Is there a reason filter_nonviable_windows isn't called here?

Reply:
I wasn't having good a result with the existing filter nonviable windows method; I'm happy to try again after making the changes that you proposed.

Reply:
It seems to be working well now after using the improved filter function.